rfcs/2020-04-20-Optimizer-minimize.md

```diff
@@ -0,0 +1,305 @@
+    self.optimizer._get_hyper(name)
+
+  def transform_loss(self, loss):
+    loss = self.optimizer.transform_loss(loss)
+    # Mixed precision needs to scale loss before calculating gradients
+    return self.scale_loss(loss)
+
+  def transform_unaggregated_gradients(self, grads_and_vars):
+    # Note: For performance, we could add a check here to see if
+    # self.optimizer.transform_unaggregated_gradients is not implemented, and if
+    # so to skip these scaling / unscalings. Or Grappler could optimize it out.
+    gradients, variables = unpack(grads_and_vars)
+    gradients = self.unscale_gradients(gradients)
+    gradients = self.optimizer.transform_unaggregated_gradients(gradients)
+    # Mixed precision needs to all-reduce on scaled gradients.
+    gradients = self.scale_gradients(gradients)
+    return zip(gradients, variables)
+
+  def aggregate_gradients(self, grads_and_vars):
+    return aggregate_in_fp16(grads_and_vars)
+
+  def transform_gradients(self, grads_and_vars):
+    gradients, variables = unpack(grads_and_vars)
+    gradients = unscale_gradients(gradients)
+    gradients = self.optimizer.transform_fgradients(gradients)
+    return zip(gradients, updates)
+
+  def apply_updates(self, grads_and_vars):
+    return self.optimizer.apply_updates(grads_and_vars)
+```
+
+(3) Horovod (only needs custom aggregation):
+
+To support backwards compatibility for Horovod:
+
+```python
+class HorovodOptimizer(Optimizer):
+  def __init__(self, optimizer):
+    self.optimizer = optimizer
+
+  def _get_hyper(self, name):
+    # See previous example
+    self.optimizer._get_hyper(name)
+
+ def aggregate_gradients(self, grads_and_vars):
+    return horovod_aggregate_gradients(grads_and_vars)
+
+ # All other methods described in this proposal simply delegate to `self.optimizer`
+```
+    
+Or, if backwards compatibility is not needed, simply:
+
+```python
+optimizer = tf.keras.optimizers.Adam(1e-3, aggregate_gradients=horovod.aggregate)
+```
+
+## `OptimizerWrapper`
+
+With this proposal, we should also release an `OptimizerWrapper` class. This class will make it easier for developers to create subclasses that wrap an `Optimizer` while providing additional functionality, such as mixed-precision, Horovod, or differential privacy use cases.
+
+## Alternatives considered
+
+#### Handle this only in Model.fit, via custom hooks exposed on the Model class
+    
+Why rejected:
+
+Shifts the responsibility for implementing and calling these hooks onto each user rather than the writer of the Optimizer subclass (Many users will write custom training logic, many fewer will write Optimizer subclasses).
+
+Solution is too Keras-specific, doesn’t solve the general problem.
+
+
+## Questions and Discussion Topics
+
+(1) Should these new hooks be private or public methods?
+(2) Should we create an initializer argument for each hook, or only for the ones we expect most users to need (`aggregate_gradients` and `transform_gradients`)?
```

Comment:
2.1 Or no constructors or no overwritables?
3. What are the contracts for the constructor arg functions? Single fns, lists, something more formal that we can better error/type-check?